dynamic-proxy/tests/common/simple_upgrade_service.rs

```diff
@@ -0,0 +1,71 @@
+use bytes::Bytes;
+use dynamic_proxy::body::{to_simple_body, SimpleBody};
+use http_body_util::{Empty, Full};
+use hyper::{
+    body::Incoming,
+    header::{HeaderValue, UPGRADE},
+    service::Service,
+    upgrade::Upgraded,
+    Request, Response, StatusCode,
+};
+use hyper_util::rt::TokioIo;
+use std::{convert::Infallible, future::Future, pin::Pin};
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+
+/// A service that upgrades the connection and echos messages back to the client.
+/// (Note: this does not use the actual websocket protocol on the wire, but is sufficient to
+/// test the upgrade path.)
+#[derive(Clone)]
+pub struct SimpleUpgradeService;
+
+impl Service<Request<Incoming>> for SimpleUpgradeService {
+    type Response = Response<SimpleBody>;
+    type Error = Infallible;
+    type Future = Pin<Box<dyn Future<Output = Result<Response<SimpleBody>, Infallible>> + Send>>;
+
+    fn call(&self, mut req: Request<Incoming>) -> Self::Future {
+        Box::pin(async move {
+            if req.headers().contains_key(UPGRADE) {
+                // Handle upgrade
+                let mut res = Response::new(to_simple_body(Empty::new()));
+                *res.status_mut() = StatusCode::SWITCHING_PROTOCOLS;
+                res.headers_mut()
+                    .insert(UPGRADE, HeaderValue::from_static("websocket"));
+
```

Comment:
@paulgb is this valid?

Reply:
It doesn't _really_ matter since we control both the client and the server here so we never actually read that header, but I guess it's nice to conform to the spec in case we use it in other tests, so I've added it.